src/query/models/tags.go

```diff
@@ -0,0 +1,250 @@
+func (t Tags) IDWithExcludes(excludeKeys ...[]byte) uint64 {
+	b := make([]byte, 0, len(t.Tags))
+	for _, tag := range t.Tags {
+		// Always exclude the metric name by default
+		if bytes.Equal(tag.Name, t.Opts.GetMetricName()) {
+			continue
+		}
+
+		found := false
+		for _, n := range excludeKeys {
+			if bytes.Equal(n, tag.Name) {
+				found = true
+				break
+			}
+		}
+
+		// Skip the key
+		if found {
+			continue
+		}
+
+		b = append(b, tag.Name...)
+		b = append(b, eq)
+		b = append(b, tag.Value...)
+		b = append(b, sep)
+	}
+
+	h := fnv.New64a()
+	h.Write(b)
+	return h.Sum64()
+}
+
+func (t Tags) tagSubset(keys [][]byte, include bool) Tags {
+	tags := NewTags(len(t.Tags), t.Opts)
+	for _, tag := range t.Tags {
+		found := false
+		for _, k := range keys {
+			if bytes.Equal(tag.Name, k) {
+				found = true
+				break
+			}
+		}
+
+		if found == include {
+			tags.AddTag(tag)
+		}
+	}
+
+	return tags
+}
+
+// TagsWithoutKeys returns only the tags which do not have the given keys.
+func (t Tags) TagsWithoutKeys(excludeKeys [][]byte) Tags {
+	return t.tagSubset(excludeKeys, false)
+}
+
+// IDWithKeys returns a string representation of the tags only including the given keys.
+func (t Tags) IDWithKeys(includeKeys ...[]byte) uint64 {
+	b := make([]byte, 0, len(t.Tags))
+	for _, tag := range t.Tags {
+		for _, k := range includeKeys {
+			if bytes.Equal(tag.Name, k) {
+				b = append(b, tag.Name...)
+				b = append(b, eq)
+				b = append(b, tag.Value...)
+				b = append(b, sep)
+				break
+			}
+		}
+	}
+
+	h := fnv.New64a()
+	h.Write(b)
+	return h.Sum64()
+}
+
+// TagsWithKeys returns only the tags which have the given keys.
+func (t Tags) TagsWithKeys(includeKeys [][]byte) Tags {
+	return t.tagSubset(includeKeys, true)
+}
+
+// WithoutName copies the tags excluding the name tag.
+func (t Tags) WithoutName() Tags {
+	return t.TagsWithoutKeys([][]byte{t.Opts.GetMetricName()})
+}
+
+// Get returns the value for the tag with the given name.
+func (t Tags) Get(key []byte) ([]byte, bool) {
+	for _, tag := range t.Tags {
+		if bytes.Equal(tag.Name, key) {
+			return tag.Value, true
+		}
+	}
+
+	return nil, false
+}
+
+// Clone returns a pointer to a copy of the tags.
+func (t Tags) Clone() *Tags {
+	// Todo: Pool these
+	clonedTags := make([]Tag, len(t.Tags))
+	copy(clonedTags, t.Tags)
+	return &Tags{
+		Tags: clonedTags,
+		Opts: t.Opts,
+	}
+}
+
+// AddTag is used to add a single tag and maintain sorted order.
+func (t *Tags) AddTag(tag Tag) Tags {
+	t.Tags = append(t.Tags, tag)
+	t.Normalize()
+	return *t
+}
+
+// SetName sets the metric name.
+func (t *Tags) SetName(value []byte) {
+	t.AddTag(Tag{Name: t.Opts.GetMetricName(), Value: value})
+}
+
+// Name gets the metric name.
+func (t Tags) Name() ([]byte, bool) {
+	return t.Get(t.Opts.GetMetricName())
+}
+
+// AddTags is used to add a list of tags and maintain sorted order.
+func (t *Tags) AddTags(tags []Tag) Tags {
```

Comment:
can we have all non pointer receivers  (or all pointers if really needed) ?

Reply:
I like the idea of not having to explicitly have stuff like `tags = tags.AddTags(fooTags)`, and rather just have `tags.AddTags(fooTags)`, with a similar approach for other add methods,

Reply:
it's error prone to have both pointer and non-pointer receivers for the same type. Can you pick one ?

Reply:
Also, it's nice to have property of Tags being immutable. Maybe you can have a builder when you need to modify and tagBuilder.Build() can normalize the values